### PR TITLE
`builtins.filter` compat with `typing.TypeGuard`

### PR DESCRIPTION
**Pull request by Jasha10**
Based on https://github.com/microsoft/pyright/issues/2768.

This change enables the following use-case:
```python
def is_not_none(x: Optional[int]) -> TypeGuard[int]:
    return x is not None

list_optional: list[Optional[int]] = [0, None, 1, None, 2]
generate_ints: Iterable[int] = filter(is_not_none, list_optional)
```

### CODE REVIEW

**Comment by Jasha10** [comment]
It has been brought to my attention that this PR is related to issue #5661 and to PR #6140.

**Comment by Jasha10** [comment]
It appears that the diff implemented in this PR is identical to that from #6140.

**Comment by JelleZijlstra** [comment]
Interesting, maybe the latest mypy release fixed things. cc @sobolevn.

**Comment by sobolevn** [comment]
Yes, this was my PR to mypy: https://github.com/python/mypy/pull/11314 after our initial attempt: #6140

**Review by sobolevn — COMMENTED** [review]
Can you please attach a simple manual test to show that it is working?

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -1052,6 +1052,8 @@ class filter(Iterator[_T], Generic[_T]):
     def __init__(self, __function: None, __iterable: Iterable[_T | None]) -> None: ...
     @overload
     def __init__(self, __function: Callable[[_T], Any], __iterable: Iterable[_T]) -> None: ...
+    @overload
+    def __init__(self, __function: Callable[[_S], TypeGuard[_T]], __iterable: Iterable[_S]) -> None: ...
```
Comment (AlexWaygood) [inline stdlib/builtins.pyi:1056]:
```suggestion
    @overload
    def __init__(self, __function: Callable[[_S], TypeGuard[_T]], __iterable: Iterable[_S]) -> None: ...
    @overload
    def __init__(self, __function: Callable[[_T], Any], __iterable: Iterable[_T]) -> None: ...
```

Here is a simple test script:

```python
from typing import Iterable, TypeGuard

def is_not_none(x: int | None) -> TypeGuard[int]:
    return x is not None

list_of_ints_or_none: list[int | None] = [0, None, 1, None, 2]
ints_list = list(filter(is_not_none, list_of_ints_or_none))
reveal_type(ints_list)
```

Run mypy on this script, with this patch applied, and the following is output (not the desired result):

```
test_filter.py:8: note: Revealed type is "builtins.list[Union[builtins.int, None]]"
```

However, if we change the order of the overloads, as I suggest here, then this is the output from mypy (the desired result):

```
test_filter.py:8: note: Revealed type is "builtins.list[builtins.int*]"
```

**Comment by hauntsaninja** [comment]
Yup, #6140 has the correct overload order. I think that wasn't tested against mypy 0.930, so worth trying again.

**Comment by sobolevn** [comment]
Let's leave this to @Jasha10, this is their first contribution 🙂

File: stdlib/builtins.pyi (hunk 1 above)
Comment (Jasha10) [reply stdlib/builtins.pyi:1056]: Thanks @AlexWaygood! I was able to reproduce the result of your test.

**Review by AlexWaygood — APPROVED** [review]

**Comment by AlexWaygood** [comment]
> Diff from [mypy_primer](https://github.com/hauntsaninja/mypy_primer), showing the effect of this PR on open source code:
> 
> ```diff
> ibis (https://github.com/ibis-project/ibis)
> + ibis/backends/csv/__init__.py:72: error: Argument 1 to "filter" has incompatible type "Callable[[Any], bool]"; expected "Callable[[Any], TypeGuard[bool]]"
> + ibis/backends/csv/__init__.py:78: error: "bool" has no attribute "name"
> + ibis/backends/csv/__init__.py:82: error: "bool" has no attribute "schema"
> + ibis/backends/csv/__init__.py:91: error: "bool" has no attribute "schema"
> + ibis/backends/csv/__init__.py:92: error: Dict entry 0 has incompatible type "bool": "Any"; expected "Node": "Any"
> ```

Huh, that's weird. I'm guessing mypy is failing to fall back to the third overload due to some weird `bool`/`TypeGuard` subtyping behaviour.

**Comment by sobolevn** [comment]
This one is the most strange: `Argument 1 to "filter" has incompatible type "Callable[[Any], bool]"; expected "Callable[[Any], TypeGuard[bool]]"`

It should just continue with other overload (probably this one: `def __init__(self, __function: Callable[[_T], Any], __iterable: Iterable[_T]) -> None: ...`). Why does it error?

I will take a look into mypy itself 👍

**Comment by sobolevn** [comment]
I am unable to reproduce this locally.
All my regular examples work as intended. But, I am using `master` branch, not `0.930`

**Comment by JelleZijlstra** [comment]
Going to merge this since the change is correct and mostly works with mypy. If it does expose a mypy bug, that can be tracked over on the mypy repo.

**Pull request merged**